proving_system/stark/src/air/example/cairo.rs

```diff
@@ -0,0 +1,305 @@
+        let last_step = self.context.trace_length - 1;
+
+        let initial_pc =
+            BoundaryConstraint::new(MEM_A_TRACE_OFFSET, 0, self.pub_inputs.pc_init.clone());
+        let initial_ap =
+            BoundaryConstraint::new(MEM_P_TRACE_OFFSET, 0, self.pub_inputs.ap_init.clone());
+
+        let final_pc = BoundaryConstraint::new(
+            MEM_A_TRACE_OFFSET,
+            last_step,
+            self.pub_inputs.pc_final.clone(),
+        );
+        let final_ap = BoundaryConstraint::new(
+            MEM_P_TRACE_OFFSET,
+            last_step,
+            self.pub_inputs.ap_final.clone(),
+        );
+
+        let constraints = vec![initial_pc, initial_ap, final_pc, final_ap];
+
+        BoundaryConstraints::from_constraints(constraints)
+    }
+
+    fn context(&self) -> AirContext {
+        self.context.clone()
+    }
+}
+
+/// From the Cairo whitepaper, section 9.10
+fn compute_instr_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    // This constraints are only applied over elements of the same row.
+    let curr = frame.get_row(0);
+
+    // Bit constraints
+    for (i, flag) in curr[0..16].iter().enumerate() {
+        constraints[i] = match i {
+            0..=14 => flag * (flag - FE::one()),
+            15 => flag.clone(),
+            _ => panic!("Unknown flag offset"),
+        };
+    }
+
+    // Instruction unpacking
+    let two = FE::from(2);
+    let b16 = two.pow(16u32);
+    let b32 = two.pow(32u32);
+    let b48 = two.pow(48u32);
+
+    let f0 = &curr[0..15]
+        .iter()
+        .rev()
+        .fold(FE::zero(), |acc, flag| flag + &two * acc);
+
+    constraints[INST] =
+        (&curr[OFF_DST]) + b16 * (&curr[OFF_OP0]) + b32 * (&curr[OFF_OP1]) + b48 * f0
+            - &curr[FRAME_INST];
+}
+
+fn compute_operand_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    // This constraints are only applied over elements of the same row.
+    let curr = frame.get_row(0);
+
+    let ap = &curr[FRAME_AP];
+    let fp = &curr[FRAME_FP];
+    let pc = &curr[FRAME_PC];
+
+    let one = FE::one();
+    let b15 = FE::from(2).pow(15u32);
+
+    constraints[DST_ADDR] =
+        &curr[F_DST_FP] * fp + (&one - &curr[F_DST_FP]) * ap + (&curr[OFF_DST] - &b15)
+            - &curr[FRAME_DST_ADDR];
+
+    constraints[OP0_ADDR] =
+        &curr[F_OP_0_FP] * fp + (&one - &curr[F_OP_0_FP]) * ap + (&curr[OFF_OP0] - &b15)
+            - &curr[FRAME_OP0_ADDR];
+
+    constraints[OP1_ADDR] = &curr[F_OP_1_VAL] * pc
+        + &curr[F_OP_1_AP] * ap
+        + &curr[F_OP_1_FP] * fp
+        + (&one - &curr[F_OP_1_VAL] - &curr[F_OP_1_AP] - &curr[F_OP_1_FP]) * &curr[FRAME_OP0]
+        + (&curr[OFF_OP1] - &b15)
+        - &curr[FRAME_OP1_ADDR];
+}
+
+fn compute_register_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    let curr = frame.get_row(0);
+    let next = frame.get_row(1);
+
+    let one = FE::one();
+    let two = FE::from(2);
+
```

Comment:
You should check the slice is big enough before starting indexing, otherwise it can panic.